lib/tbot/tbot_test.go

```diff
@@ -0,0 +1,196 @@
+	// is resolved.
+	fc.Databases.EnabledFlag = "false"
+
+	cfg := service.MakeDefaultConfig()
+	require.NoError(t, config.ApplyFileConfig(fc, cfg))
+	cfg.FileDescriptors = fds
+	cfg.Log = log
+	cfg.CachePolicy.Enabled = false
+	cfg.Proxy.DisableWebInterface = true
+
+	svcC := make(chan *service.TeleportProcess)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := service.Run(ctx, *cfg, func(cfg *service.Config) (service.Process, error) {
+			svc, err := service.NewTeleport(cfg)
+			if err == nil {
+				svcC <- svc
+			}
+			return svc, err
+		})
+		require.NoError(t, err)
+	}()
+
+	var svc *service.TeleportProcess
+	select {
+	case <-time.After(10 * time.Second):
+		t.Fatal("teleport process did not instantiate in 10 seconds")
+	case svc = <-svcC:
+	}
+
+	// Ensure the service starts correctly the first time before proceeding
+	eventCh := make(chan service.Event, 1)
+	svc.WaitForEvent(svc.ExitContext(), service.TeleportReadyEvent, eventCh)
+	select {
+	case <-eventCh:
+	case <-time.After(30 * time.Second):
+		// in reality, the auth server should start *much* sooner than this.  we use a very large
+		// timeout here because this isn't the kind of problem that this test is meant to catch.
+		t.Fatal("auth server didn't start after 30s")
+	}
+
+	// Tracks the latest instance of the Teleport service through reloads
+	activeSvc := svc
+	activeSvcMu := sync.Mutex{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case svc := <-svcC:
+				activeSvcMu.Lock()
+				activeSvc = svc
+				activeSvcMu.Unlock()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return testhelpers.MakeDefaultAuthClient(t, log, fc), func() *service.TeleportProcess {
+		activeSvcMu.Lock()
+		defer activeSvcMu.Unlock()
+		return activeSvc
+	}, fc
+}
+
+// TestCARotation is a heavy integration test that through a rotation, the bot
+// receives credentials for a new CA.
+func TestBot_Run_CARotation(t *testing.T) {
+	t.Parallel()
+	if testing.Short() {
+		t.Skip("test skipped when -short provided")
+	}
+
+	// wg and context manage the cancellation of long running processes e.g
+	// teleport and tbot in the test.
+	log := libutils.NewLoggerForTests()
+	wg := &sync.WaitGroup{}
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(func() {
+		log.Infof("Shutting down long running test processes..")
+		cancel()
+		wg.Wait()
+	})
+
+	client, teleportProcess, fc := setupServerForCARotationTest(ctx, log, t, wg)
+
+	// Make and join a new bot instance.
+	botParams := testhelpers.MakeBot(t, client, "test", "access")
+	botConfig := testhelpers.MakeMemoryBotConfig(t, fc, botParams)
+	b := New(botConfig, log, make(chan struct{}))
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := b.Run(ctx)
+		require.NoError(t, err)
+	}()
+	// Allow time for bot to start running and watching for CA rotations
+	// TODO: We should modify the bot to emit events that may be useful...
+	time.Sleep(10 * time.Second)
```

Comment:
How do we know that all of these sleeps will be long enough to be reliable and not flaky?

Reply:
Being frank, we don't. It's a bit rubbish, but I was worried about adding more complexity to this PR (in the form of some kind of event emitter) purely to service the tests.  Ten seconds is very generous here for the amount of things it needs to set up though.

Whilst a bit nebulous, it is my full intention to get around to adding this sort of "Ready" event emitter. One thing I have in mind for Machine ID is potentially a model where user's writing Go would be able to directly embed it within their codebase, and this sort of ready event would be needed by them.